### PR TITLE
[Post] Content-Visibility

### LINKED ISSUE #3543: content: CSS Content Visibility [2020-07-28]
* [draft](https://docs.google.com/document/d/1hLT9of1rENxejNgyZW8E0X4k547nhkzsKMK7U_owCo0/edit#heading=h.skawfgvytwwk)
* [thread](https://groups.google.com/a/google.com/g/webdev-content/c/0Sd37wv1DUE)
* next step: @kaycebasques to review

history:

* @kaycebasques reviewed on 7/24

### PR DESCRIPTION
**Pull request by una**
Resolves #3543 by adding post about content visibility

### CODE REVIEW

File: src/site/content/en/blog/content-visibility/index.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
+---
+title: 'CSS Content Visibility: the most impactful performance boost CSS can give you'
```
Comment (kaycebasques) [inline src/site/content/en/blog/content-visibility/index.md:2]: @una are you sure about the statement that this is the most impactful performance boost? If not there are ways for us to rephrase this to avoid a false claim

File: src/site/content/en/blog/content-visibility/index.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  - css
+---
+
+The [`content-visibility`](https://drafts.csswg.org/css-contain/#propdef-content-visibility) property might be one of the most impactful new CSS properties for improving page load performance. `content-visibility` enables the user agent to skip an element's rendering work, including layout and painting, until it is needed. If a large portion of your content is off-screen, leveraging the `content-visibility` property allows the browser to skip the rendering working for this content, making the initial user load much faster. Additionally, it allows for faster interactions with the on-screen content. Pretty neat.
+
+`content-visibility` relies on primitives within the [the CSS Containment Spec](http://drafts.csswg.org/css-contain/). So, what exactly is this spec? Let's dive in.
+
+## CSS Containment {: #containment }
+
+ -- Caniuse Image Here: caniuse.com/#feat=css-containment --
```
Comment (kaycebasques) [inline src/site/content/en/blog/content-visibility/index.md:22]:
```suggestion
`content-visibility` relies on primitives within the [the CSS Containment Spec](http://drafts.csswg.org/css-contain/). So, what exactly is this spec? Let's dive in.

## Browser compatibility

 -- Caniuse Image Here: caniuse.com/#feat=css-containment --

## CSS Containment {: #containment }

```

I prefer to have the compat info in a separate section so that if someone scans the article it's very easy to find

File: src/site/content/en/blog/content-visibility/index.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+The key and overarching goal of CSS containment is to enable rendering performance improvements of web content by providing *predictable isolation of a DOM subtree* from the rest of the page.
```
Comment (kaycebasques) [inline src/site/content/en/blog/content-visibility/index.md:24]:
```suggestion
The key and overarching goal of CSS containment is to enable rendering performance improvements of web content by providing **predictable isolation of a DOM subtree** from the rest of the page.
```
I think bold gets the emphasis across a bit more than italics. WDYT?

File: src/site/content/en/blog/content-visibility/index.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+There are four types of [CSS containment](https://developers.google.com/web/updates/2016/06/css-containment), each a potential value for the `contain` CSS property. You can also combine them individually or together in a space-separated list of values:
+
+- `size`: Size containment ensures that the containing box can be laid out without needing to examine its children. The parent sets the geometric size independent of its contents.
+- `layout`: Layout containment means that the subtree does not affect the external layout of other boxes on the page.
+- `style`: Style containment ensures that properties which can have effects on more than just its descendants don't escape the element (e.g. counters).
+- `paint`: Paint containment ensures that the descendants of the containing box don't display outside its bounds. Nothing can visibly overflow the element, and if an element is off-screen or otherwise not visible, its descendants will also not be visible.
+
+## `content-visibility`
```
Comment (kaycebasques) [inline src/site/content/en/blog/content-visibility/index.md:35]:
```suggestion
## Skipping rendering work with `content-visibility`
```
@una I updated each of these sections to use the format "X with Y" where X is a generic description of the task and Y is the property name. I feel that this is easier to scan and is less intimidating for people who are learning about these concepts for the first time. WDYT? Feel free to rephrase any of them.

File: src/site/content/en/blog/content-visibility/index.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+It may be hard to figure out which containment values to use, since browser optimizations may only kick in when an appropriate set is specified. You can play around with the values to see [what works best](https://developers.google.com/web/updates/2016/06/css-containment), or you can use another CSS property called `content-visibility` to apply the needed containment automatically. `content-visibility` was created to ensure that you get the largest performance gains the browser can provide with least effort. 
+
+### `content-visibility: auto` {: #auto }
```
Comment (kaycebasques) [inline src/site/content/en/blog/content-visibility/index.md:39]:
```suggestion
### Automatic performance improvements with `content-visibility: auto` {: #auto }
```

File: src/site/content/en/blog/content-visibility/index.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  content-visibility: auto;
+  contain-intrinsic-size: 100px 500px; /* Explained in the next section. */
+}
+```
+
+{% Aside %}
+Note that as content moves in and out of visibility, it will start and stop being rendered as needed. However, this does not mean that the browser will have to render and re-render the same content over and over again.
+{% endAside %}
+
+##### `contain-intrinsic-size`
```
Comment (kaycebasques) [inline src/site/content/en/blog/content-visibility/index.md:83]:
```suggestion
##### Specifying the natural size of an element with `contain-intrinsic-size`
```

File: src/site/content/en/blog/content-visibility/index.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+This might not be ideal, since the size of the scrollbar will shift, being reliant on each story having a non-zero height.
+
+Thankfully, CSS provides another property, `contain-intrinsic-size`, which effectively specifies the natural size of the element *if the element is affected by size containment*.
+
+This means it will lay out as if it had a single child of "intrinsic-size" dimensions, ensuring that your unsized divs still occupy space. `contain-intrinsic-size` acts as a placeholder size in lieu of rendered content.
+
+
+
+### `content-visibility: hidden`
```
Comment (kaycebasques) [inline src/site/content/en/blog/content-visibility/index.md:97]:
```suggestion
### Hiding content with `content-visibility: hidden`
```

File: src/site/content/en/blog/content-visibility/index.md [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+authors:
+  - vladimirlevin
+  - una
+date: 2020-08-05
+hero: hero.jpg
+alt: Stylized photo of a half-hidden person.
+description: TBD
+tags:
+  - blog
+  - css
```
Comment (kaycebasques) [inline src/site/content/en/blog/content-visibility/index.md:13]:
```suggestion
  - css
  - layout
```

File: src/site/content/en/blog/content-visibility/index.md (hunk 4 above)
Comment (una) [reply src/site/content/en/blog/content-visibility/index.md:35]: I like it!

File: src/site/content/en/blog/content-visibility/index.md (hunk 1 above)
Comment (una) [reply src/site/content/en/blog/content-visibility/index.md:2]: We are renaming it to: `content-visibility`: the new CSS property that boosts your rendering performance

**Comment by una** [comment]
Hey @kaycebasques -- is this good to go now?
